projects/01-rubygems-cli/solution/lib/ruby_gems_api.rb

```diff
@@ -0,0 +1,33 @@
+require "faraday"
+require "faraday/net_http"
+require "json"
+require "./errors/gem_not_found"
+require "./errors/bad_request"
+
+class RubyGemsApi
+  @connection = Faraday.new("https://rubygems.org") do |faraday|
+    faraday.adapter :net_http
+  end
+
+  class << self
+    def gem_info(gem_name)
+      raise ArgumentError.new("Invalid gem name provided.") if gem_name.nil? || gem_name.empty?
+      
+      response = @connection.get("/api/v1/gems/#{gem_name}.json")
+      raise GemNotFound.new("Gem #{gem_name} not found.") if response.body == "This rubygem could not be found."
+      
+      raise BadRequest.new("Bad request.") if response.status != 200
+      
+      JSON.parse(response.body)
+    end
+
+    def search_gems(keyword)
+      raise ArgumentError.new("Invalid keyword provided.") if keyword.nil? || keyword.empty?
+
+      response = @connection.get("/api/v1/search.json", {query: keyword})
+      raise BadRequest.new("Bad request.") if response.status != 200
```

Comment:
BadRequest maps to the HTTP 400 (Bad Request) code. I'd name this error more generic, because BadRequest means that the client did something wrong, and here it may be the case where the API is inaccessible (HTTP 5xx code).

Reply:
f993fbfd442105c4af595e6803be0c8ec5142119